File: lib/net-abort.js
var net = require('net');
var net_abort = require('../build/Release/net_abort');


/**
 * Emit an RST packet and close socket.
 * @param {net.Socket} socket - an connected TCP socket to close.
 */
function abort() {
  var socket;
  if (this instanceof net.Socket) {
    socket = this;
  } else if (arguments[0] instanceof net.Socket) {
    socket = arguments[0];
  } else {
    throw new TypeError('Supplied argument is not net.Socket');
  }
  
  if (!socket._handle) {
    throw new Error('Socket is not connected or is not TCP socket');
  }
  
  net_abort.abort.apply(socket._handle);
  socket.destroy();
}

/**
 * Install socket.abort to net.Socket class
 */
function install() {
  net.Socket.prototype.abort = abort;
}

module.exports = abort;
module.exports.install = install;

File: src/net_abort.cc
#include <node.h>
#include <tcp_wrap.h>
#include <v8.h>
#include <nan.h>

#ifdef WIN32
	#include <winsock2.h>
#else
	#include <unistd.h>
	#include <errno.h>
#endif


using namespace node;
using namespace v8;


/**
 * Emit an RST packet and close socket.
 * @this TCP
 */
NAN_METHOD(Abort) {
	NanScope();

	// TODO: type check
	TCPWrap* wrap = (TCPWrap *)NanGetInternalFieldPointer(args.Holder(), 0);

#ifdef WIN32
	SOCKET socket = wrap->UVHandle()->socket;
	linger val = { 1, 0 };
	if (setsockopt(socket, SOL_SOCKET, SO_LINGER, (char *)&val, sizeof(val)) != 0) {
		NanThrowError(WinapiErrnoException(WSAGetLastError(), "setsockopt"));
		NanReturnUndefined();
	}
	if (closesocket(socket) != 0) {
		NanThrowError(WinapiErrnoException(WSAGetLastError(), "closesocket"));
		NanReturnUndefined();
	}
#else
	int fd = wrap->UVHandle()->io_watcher.fd;
	linger val = { 1, 0 };
	if (setsockopt(fd, SOL_SOCKET, SO_LINGER, (char *)&val, sizeof(val)) != 0) {
		NanThrowError(ErrnoException(errno, "setsockopt"));
		NanReturnUndefined();
	}
	if (close(fd) != 0) {
		NanThrowError(ErrnoException(errno, "close"));
		NanReturnUndefined();
	}
#endif

	NanReturnUndefined();
}

/**
 * Initialize this module
 * @param {Handle<Object>} exports - Module object
 */
void init(Handle<Object> exports) {
	exports->Set(NanNew<String>("abort"), NanNew<FunctionTemplate>(Abort)->GetFunction());
}

// HACK: no TCPWrap::UVHandle in node.lib
uv_tcp_t* TCPWrap::UVHandle() {
	return &handle_;
}

/**
 * Module declaration
 */
NODE_MODULE(net_abort, init);
